sync/backup.go

```diff
@@ -150,16 +153,99 @@ func (t *Task) restoreBackup(replicaInController *types.ControllerReplicaInfo, b
 	if err != nil {
 		return err
 	}
+	credential := map[string]string{}
+	backupType, err := util.CheckBackupType(backup)
+	if err != nil {
+		return err
+	}
+	if backupType == "s3" {
+		accessKey := os.Getenv(types.AWSAccessKey)
+		if accessKey == "" {
+			return fmt.Errorf("Missing environment variable AWS_ACCESS_KEY_ID for s3 backup")
+		}
+		secretKey := os.Getenv(types.AWSSecretKey)
+		if secretKey == "" {
+			return fmt.Errorf("Missing environment variable AWS_SECRET_ACCESS_KEY for s3 backup")
+		}
+		credential[types.AWSAccessKey] = accessKey
+		credential[types.AWSSecretKey] = secretKey
+		credential[types.AWSEndPoint] = os.Getenv(types.AWSEndPoint)
+	}
 
 	logrus.Infof("Restoring backup %s on %s", backup, replicaInController.Address)
 
-	if err := repClient.RestoreBackup(backup, snapshotFile); err != nil {
+	if err := repClient.RestoreBackup(backup, snapshotFile, credential); err != nil {
 		logrus.Errorf("Failed restoring backup %s on %s", backup, replicaInController.Address)
 		return err
 	}
+
 	return nil
 }
 
+func (*Task) IsRestorationComplete(rsList []*RestoreStatus) bool {
+	for _, r := range rsList {
+		if r.Progress != 100 {
+			return false
+		}
+	}
+	return true
+}
+
+func (t *Task) GetRestoreStatus() ([]*RestoreStatus, error) {
+	var restoreStatusList []*RestoreStatus
+	snapshotFile := ""
+	replicas, err := t.client.ReplicaList()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, replica := range replicas {
+		repClient, err := replicaClient.NewReplicaClient(replica.Address)
+		if err != nil {
+			logrus.Errorf("Cannot create a replica client for IP[%v]: %v", replica.Address, err)
+			return nil, err
+		}
+
+		rsList, err := repClient.BackupRestoreStatus()
+		if err != nil {
+			return nil, err
+		}
+
+		if rsList.DestFileName == "" {
+			//No Active Backup Restore happening
+			continue
+		}
+		snapshotFile = rsList.DestFileName
+
+		restoreStatusList = append(restoreStatusList, &RestoreStatus{
+			Progress:     int(rsList.Progress),
+			RestoreError: rsList.Error,
+			SnapshotName: rsList.DestFileName,
+		})
+	}
+
+	if len(restoreStatusList) == 0 {
+		//No Active Backup Restore happening
+		return nil, nil
+	}
+
+	if t.IsRestorationComplete(restoreStatusList) {
+		snapshotID, err := replica.GetSnapshotNameFromDiskName(snapshotFile)
+		if err != nil {
+			return nil, err
+		}
+
+		// call to controller to revert to sfile
+		if err := t.client.VolumeRevert(snapshotID); err != nil {
+			return nil, fmt.Errorf("Fail to revert to snapshot %v: %v", snapshotID, err)
+		}
+
+		//TODO: Call for sync restore object delete
```

Comment:
We can leave the previous restoration information there. I don't think we need to hurry to delete it since if later it happens again, the old one will be overridden.